readthedocs/donate/models.py

```diff
@@ -50,9 +59,48 @@ def __str__(self):
 
     def as_dict(self):
         "A dict respresentation of this for JSON encoding"
+        image_url = reverse(
+            'donate_view_proxy',
+            kwargs={'promo_id': self.pk, 'hash': random.randint(0, 10000000)}
+        )
+        link_url = reverse(
+            'donate_click_proxy',
+            kwargs={'promo_id': self.pk}
+        )
         return {
             'id': self.analytics_id,
             'text': self.text,
-            'link': self.link,
-            'image': self.image,
+            'link': link_url,
+            'image': image_url,
         }
+
+    def incr(self, type):
+        assert type in ['offers', 'views', 'clicks']
+        day = get_ad_day()
+        impression, _ = self.impressions.get_or_create(date=day)
+        setattr(impression, type, models.F(type) + 1)
+        impression.save()
+
+    def incr_redis(self, type):
+        assert type in ['offers', 'views', 'clicks']
+        day = get_ad_day()
+        redis = Redis.from_url(settings.BROKER_URL)
+        redis.incr('{slug}-{year}-{month}-{day}-{type}'.format(
```

Comment:
A couple things here:

* This is too general of a namespace, it should signify this is a promo event in the namespace
* Type should be in the front of the key name, as we'll want to do filtering on that primarily
* This might lend to using a redis hash increment [1] instead of just an increment, for cleanliness -- that is:

```python
redis.hincrby('promo:{slug}:{type}:daily'.format(...), '{year}-{month}-{day}'.format(...), 1)
```

Using a hash increment and a key with `type` moved forward, it's then easier to get the values out:

```python
values = redis.hvals('promo:{slug}:{type}:daily'.format(...))
sum(values)
```

* Lastly, it would make sense to also increment broader keys, such as:

```python
redis.hincrby('promo:{slug}:{type}:daily'.format(...), '{year}-{month}-{day}'.format(...), 1)
redis.hincrby('promo:{slug}:{type}:monthly'.format(...), '{year}-{month}'.format(...), 1)
redis.incr('promo:{slug}:{type}'.format(...))
```

1: http://redis.io/commands/hincrby